tests/test_cache.py

```diff
@@ -0,0 +1,105 @@
+):
+    return redis.ConnectionPool(host=redis_host, port=redis_port, db=redis_db)
+
+
+class TestCachedSession(unittest.TestCase):
+    def tearDown(self):
+        try:
+            shutil.rmtree(file_cache_directory)
+        except:
+            pass
+
+    @httpretty.activate
+    def test_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+        )
+        session = CachedSession(
+            fallback_cache_duration=10,
+            file_cache_directory=file_cache_directory,
+        )
+        resp = session.get("https://httpbin.org/ip")
+
+        for h in resp.headers:
+            if h == "cache-control":
+                print(resp.headers[h])
+
+        self.assertIn(
+            "Automatically cached for 0:00:10", resp.headers["warning"]
+        )
+
+    @httpretty.activate
+    def test_default_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertIn(
+            "Automatically cached for 0:00:05", resp.headers["warning"]
+        )
+
+    @httpretty.activate
+    def test_cache_control_max_age_overwrites_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+            adding_headers={"Cache-Control": "max-age=30"},
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertNotIn("warning", resp.headers)
+        self.assertIn("max-age=30", resp.headers["cache-control"])
+
+    @httpretty.activate
+    def test_cache_control_no_cache_overwrites_custom_heuristic(self):
+        httpretty.register_uri(
+            httpretty.GET,
+            "https://httpbin.org/ip",
+            body='{"origin": "127.0.0.1"}',
+            adding_headers={"Cache-Control": "no-cache"},
+        )
+        session = CachedSession(file_cache_directory=file_cache_directory)
+        resp = session.get("https://httpbin.org/ip")
+
+        self.assertNotIn("warning", resp.headers)
+        self.assertIn("no-cache", resp.headers["cache-control"])
+
+    def test_file_cache(self):
+        session = CachedSession()
+        adapter = session.get_adapter("http://")
+        cache = adapter.cache
+        self.assertIsInstance(cache, FileCache)
+
+    def test_redis_cache(self):
+        pool = create_redis_connection_pool()
+        session = CachedSession(redis_cache_pool=pool)
+        adapter = session.get_adapter("http://")
+        cache = adapter.cache
+        self.assertIsInstance(cache, RedisCache)
```

Comment:
Same here - what we need to test is that the module can successfully connect to and store information in a redis store (or a mock redis store), and retrieve that information again.

Checking that a `RedisCache` instance is attached to an adapter doesn't mean it's being used, and conversely the lack of the adapter doesn't mean it's not using Redis through some other means.

I imagine properly testing a Redis connection will be quite tricky though, so I'd be happy for this PR to be merged without this test for now - we can verify Redis manually. Unless you particularly want to implement the test properly.

Reply:
Implemented it with a **redis-mock** library. One thing I had to add is a wrapper that ensures that the mocked redis object is a singleton for all consumers. This allows the manipulation of the cache entries from outside the library. 

I still have the same problem here though. The library could encode the cached response in any way. Ive tried to directly manipulate the data at a byte level here. 
This should be fairly flexible.